libcudacxx/include/cuda/std/detail/libcxx/include/__functional/reference_wrapper.h

```diff
@@ -60,7 +60,7 @@ class _LIBCUDACXX_TEMPLATE_VIS reference_wrapper : public __weak_result_type<_Tp
 
 #if _LIBCUDACXX_STD_VER > 14 && !defined(_LIBCUDACXX_HAS_NO_DEDUCTION_GUIDES)
 template <class _Tp>
-reference_wrapper(_Tp&) -> reference_wrapper<_Tp>;
+_LIBCUDACXX_HOST_DEVICE reference_wrapper(_Tp&) -> reference_wrapper<_Tp>;
```

Comment:
@Artem-B I am really wondering whether this is an intentional choice of clang-cuda or just a consequence of how deduction guides are implemented.

Currently we need to add `__host__ __device__` to them . I do not really see a use case because dispatching to different types via deduction guides seems super icky

Reply:
It may be similar to https://github.com/llvm/llvm-project/issues/67507, only there it's an interaction w/ concepts.

@yxsamliu -- ^^^ FYI

Attribute-based overload resolution requires correct `caller` context and some of the newer C++ features may not have them plumbed through yet. In some cases, we don't really know yet how to make such overloading and C++ interact in a consistent way. 

E.g. if the source code happens to have both a `__device__ reference_wrapper()` and a `__host__ reference_wrapper()`, which one the deduction guide has in mind? 

In this case, overload resolution happens without the caller context and that may not work for a host-only function during GPU-side compilation. Making the function HD makes it available in both host and GPU contexts, so it works around the issue.

Reply:
The current behaviour seems reasonable regarding host/device attributes of the deduction guide. Basically, host/device attribute of a deduction guide means it can be used in the corresponding context, which is similar to the meaning of host/device attribute of a function declaration. By default, it assumes host attribute. This is consistent with C++.

However, there seems to be an issue with the implicit deduction guide https://godbolt.org/z/4zsvrf8K1

It seems the implicit deduction guide by default is host device, which causes ambiguity when there is no explicit deduction guide or the explicit deduction guide cannot be used due to host/device attributes. If the implicit deduction guide takes the same host/device attribute of the ctor, the issue should disappear.